### ISSUE TITLE
[Bug] Using q instead of normalized q in Megatron's DSA MLA indexer for GLM 5 models

### ISSUE THREAD

**Issue by SamuelGabriel** [THUDM/slime#2165]
### Bug Description

In `slime_plugins/models/glm5/glm5.py`, the DSA lightning-indexer projects its query from the **raw** `q_a_proj` output (`q_compressed`), i.e. it skips the `q_a_layernorm` RMSNorm. The reference GLM-5 / DeepSeek-V3.2 implementations (HF `transformers` and vLLM) feed the indexer the `q_a_layernorm`-normalized q. 

I've run an experiment to see the difference of doing this or not and couldn't see a significant difference in terms of abs diff between the logprobs of vLLM and megatron, but wanted to raise this still, since I believe this should actually change the top k and thus make the model a little bit off policy.

## Locations in the code

I've let AI generate views of where this happens (or rather not) in slime vs HF reference to find it easily in the code:

### Where slime skips the norm

`q_layernorm` is `IdentityOp`, so the q-a-layernorm is fused only into the *main* `linear_q_up_proj`; the indexer reads the pre-norm `q_compressed`:

```python
# slime_plugins/models/glm5/glm5.py
q_compressed, _ = self.linear_q_down_proj(hidden_states)   # L537: raw q_a_proj output
...
q_compressed = self.q_layernorm(q_compressed)              # L551: q_layernorm = IdentityOp -> no-op
q, _ = self.linear_q_up_proj(q_compressed)                 # L552: main path normalizes internally (fused LN)
...
q_compressed = q_compressed.detach()                       # L616: still un-normalized
index_q, _ = self.wq_b(q_compressed)                       # L620: indexer query from RAW q
```
- https://github.com/THUDM/slime/blob/fa3c990af6f18efd3fd9922698bf4bf4048d1263/slime_plugins/models/glm5/glm5.py#L616-L620
- spec sets `q_layernorm=IdentityOp`: https://github.com/THUDM/slime/blob/fa3c990af6f18efd3fd9922698bf4bf4048d1263/slime_plugins/models/glm5/glm5.py#L764

### Reference (HF `transformers`)

`GlmMoeDsaAttention` computes `q_resid = q_a_layernorm(q_a_proj(hidden))` and passes **that** to the indexer, whose `wq_b` consumes it:

```python
# transformers/models/glm_moe_dsa/modeling_glm_moe_dsa.py
q_resid = self.q_a_layernorm(self.q_a_proj(hidden_states))   # L423
...
topk_indices = self.indexer(hidden_states, q_resid, ...)     # L446

# class GlmMoeDsaIndexer.forward:
#   q_resid: Query residual from `q_a_layernorm(q_a_proj(x))`  (L220, docstring)
q = self.wq_b(q_resid)                                       # L231
```
- https://github.com/huggingface/transformers/blob/8edd87b6331bf4f6fb290a6ee302637574ec2e7d/src/transformers/models/glm_moe_dsa/modeling_glm_moe_dsa.py#L423
- https://github.com/huggingface/transformers/blob/8edd87b6331bf4f6fb290a6ee302637574ec2e7d/src/transformers/models/glm_moe_dsa/modeling_glm_moe_dsa.py#L446
- https://github.com/huggingface/transformers/blob/8edd87b6331bf4f6fb290a6ee302637574ec2e7d/src/transformers/models/glm_moe_dsa/modeling_glm_moe_dsa.py#L220-L231

### Steps to Reproduce

Just any normal GLM 5 training.

### Expected Behavior

We are very close to expected behavior, this is more of a slight imprecision.

### Actual Behavior

No normalization, where HF normalizes.

### Environment

none

### Logs

```shell

```

### Pre-submission Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/THUDM/slime/blob/main/CONTRIBUTING.md) and understand the collaboration scope.
- [x] I have read the [documentation](https://thudm.github.io/slime/) and my issue is not addressed there.
- [x] I have searched for [existing issues](https://github.com/THUDM/slime/issues) and this is not a duplicate.
- [x] I have provided a minimal, reproducible example.

**Comment by n-WN** [comment]
I'd like to take this issue. I traced the path on current `main` and confirmed the report, then ran a weight-level experiment to judge whether it actually matters — since, as you observed, logprob abs diffs look negligible. Sharing the reasoning and data below.

## Confirming the report

- In `get_glm5_spec`, `q_layernorm=IdentityOp` and the q_a_layernorm is fused into `linear_q_up_proj` (`column_parallel_layer_norm_linear`), so at the indexer block `q_compressed` is still the raw `q_a_proj` output when `wq_b` consumes it (`slime_plugins/models/glm5/glm5.py`).
- Both reference implementations feed the indexer the **normalized** q:
  - HF `modeling_glm_moe_dsa.py`: `q_resid = self.q_a_layernorm(self.q_a_proj(hidden_states))` → `self.indexer(..., q_resid, ...)` (the docstring also states `q_resid` comes from `q_a_layernorm`);
  - SGLang `models/deepseek_common/attention_forward_methods/forward_mla.py`: `q = self.q_a_layernorm(q)` → `q_lora = q` (`# q_lora needed by indexer`) → `self.indexer(q_lora, ...)`.

  Since SGLang is the rollout engine, this is a train/rollout mismatch in the DSA top-k selection.

## Why logprob diffs hide it — and what the real effect is

The index score is `Σ_h w_h · relu(q_h · k · scale)` followed by top-k. `relu` is positively homogeneous, so a **per-token positive scalar** on q scales the whole score row uniformly and provably cannot change the top-k. RMSNorm is `(x / rms(x)) ⊙ g`: the `1/rms` part is exactly such a scalar, so **the entire effect of skipping the norm reduces to the per-dim gain `g`**.

I verified this with real GLM-5.2 checkpoint tensors (`q_a_proj`, `q_a_layernorm`, `indexer.{wq_b, wk, weights_proj, k_norm}` for computing layers 2 / 38 / 66), using the exact HF index-score formula:

- **Control**: rms-only normalization without the gain gives top-k overlap **1.000000** vs raw q — validating both the derivation and the simulation.
- **The real gains are far from identity**: e.g. layer 2 has mean 0.148 / min 0.041 / max 0.316; several layers contain near-zero and even **negative** gain entries (sign flips), so this is not a benign rescale.
- **Top-2048 overlap** between raw-q selection (current training side) and normalized-q selection (rollout side), N=8192, positions ≥ 4096, stable across two seeds:

  | layer | mean overlap | p10 | worst |
  |---|---|---|---|
  | 2 | 0.90 | 0.86 | 0.79 |
  | 38 | 0.95 | 0.89 | 0.82 |
  | 66 | 0.90 | 0.83 | 0.79 |

So on average ~5–10% (tail positions: up to ~20%) of the selected KV positions differ from what the rollout engine used to generate the data — a structural off-policy effect on the sparse attention pattern that per-token logprob diffs mostly average away. (Caveat: hidden states are synthetic Gaussians with the q covariance inherited from the real `q_a_proj`; real activations may shift the numbers somewhat, but the control isolates the mechanism and the gain spread is a property of the released weights.)

## Proposed fix

Recompute the norm for the indexer query from `linear_q_up_proj.layer_norm_weight` (which maps to HF `q_a_layernorm.weight`, see `megatron_to_hf/deepseekv3.py`), exactly mirroring how the same function already recomputes the fused kv_a_layernorm for `kv_compressed` via `torch.nn.functional.rms_norm`. Note the indexer receives no gradients on the training side (top-k scores are discarded, indices are integer), so this is a pure forward/parity change with no training-dynamics side effects.

I have the fix and the experiment ready; PR incoming if no one else is already on this.
